### PR TITLE
Add `security.protection.start` to optionally prevent instance start up

### LINKED ISSUE #12110: Protect/Lock VMs or Containers from starting
# Required information

-- Feature Request --

 * Distribution: Ubuntu
 * Distribution version: 2204
 * The output of "lxc info" or if that fails:
   * LXD version: 5 Stable
   * Storage backend in use: ZFS Local storage

# Issue description

I want to create VM templates with Windows and Linux and lock them to prevent starting. It will mess up my sysprep configuration. Think of lxc start --all.  It is not (yet) possible to create or convert VMs to templates in LXD. For now, a lock that prevents starting would suffice.

https://discourse.ubuntu.com/t/is-it-possible-to-lock-an-instance/37496/5?u=tomvb

# Steps to reproduce

 1. Step one - Create VM with Windows and sysprep the VM
 2. Step two - lxc start VM name
 3. Step three - Oops, i dont want to start the VM with sysprep, I want to clone the VM with sysprep.

### PR DESCRIPTION
**Pull request by qianzhangxa**
This PR adds a new instance configuration option `security.protection.start`, and throws an error if a given instance with this option set to `true` is started.

Closes #12110.

### CODE REVIEW

**Comment by tomponline** [comment]
Hi please can you rebase to fix conflicts

**Comment by qianzhangxa** [comment]
> Hi please can you rebase to fix conflicts

@tomponline Done, thanks.

File: shared/version/api.go [hunk 1]
```diff
@@ -411,6 +411,7 @@ var APIExtensions = []string{
 	"shared_custom_block_volumes",
 	"instance_import_conversion",
 	"instance_create_start",
+	"container_protection_start",
```
Comment (tomponline) [inline shared/version/api.go:414]: should be `instance_protection_start` as it covers VMs as well

File: test/suites/security.sh [hunk 2]
```diff
@@ -161,6 +161,22 @@ test_security_protection() {
 
   lxc profile unset default security.protection.delete
 
+  # Test start protection
+  lxc init testimage c1
```
Comment (tomponline) [inline test/suites/security.sh:165]: I dont think we need these 3 steps as we have plenty of container start tests :)

**Review by tomponline — CHANGES_REQUESTED** [review]
I think `instanceShouldAutoStart` should be updated to take into account this new setting to avoid repeated start up tries and errors.

**Comment by tomponline** [comment]
Please can you sign the Canonical CLA too. https://ubuntu.com/legal/contributors Thanks

File: test/suites/security.sh (hunk 2 above)
Comment (qianzhangxa) [reply test/suites/security.sh:165]: Sorry, which 3 steps? Did you mean the 3 steps below?
```
  lxc init testimage c1
  lxc start c1
  lxc delete c1 --force
```

File: shared/version/api.go (hunk 1 above)
Comment (qianzhangxa) [reply shared/version/api.go:414]: Agree.

And how about `container_protection_delete`? I think it has the same problem, do we need to rename it?

**Comment by qianzhangxa** [comment]
> Please can you sign the Canonical CLA too. https://ubuntu.com/legal/contributors Thanks

I have signed it just now. Can we rerun the CLA check job to verify?

image

File: shared/version/api.go (hunk 1 above)
Comment (tomponline) [reply shared/version/api.go:414]: No, because that would break anyone checking for it.

File: test/suites/security.sh (hunk 2 above)
Comment (tomponline) [reply test/suites/security.sh:165]: yes, this doesnt test anything we dont already test

**Comment by tomponline** [comment]
> > Please can you sign the Canonical CLA too. https://ubuntu.com/legal/contributors Thanks
> 
> I have signed it just now. Can we rerun the CLA check job to verify?
> 
> image

Thanks will ask for a CLA list refresh internally....

**Review by tomponline — CHANGES_REQUESTED** [review]
I think instanceShouldAutoStart should be updated to take into account this new setting to avoid repeated start up tries and errors.

File: shared/version/api.go (hunk 1 above)
Comment (qianzhangxa) [reply shared/version/api.go]: Done.

File: test/suites/security.sh (hunk 2 above)
Comment (qianzhangxa) [reply test/suites/security.sh]: Done.

**Comment by qianzhangxa** [comment]
> I think instanceShouldAutoStart should be updated to take into account this new setting to avoid repeated start up tries and errors.

@tomponline I have updated `instanceShouldAutoStart` in the latest push, my solution is to check `security.protection.start` first and then `boot.autostart` (i.e. the former has higher priority than the latter), please let me know if you have a different idea, thanks!

**Comment by qianzhangxa** [comment]
@tomponline I am thinking about this case: If we set `security.protection.start` to `true` for a **running** instance, do we need to automatically stop the instance?

**Comment by tomponline** [comment]
> @tomponline I have updated `instanceShouldAutoStart` in the latest push, my solution is to check `security.protection.start` first and then `boot.autostart` (i.e. the former has higher priority than the latter), please let me know if you have a different idea, thanks!

Sounds good! Thanks

**Comment by tomponline** [comment]
> @tomponline I am thinking about this case: If we set `security.protection.start` to `true` for a **running** instance, do we need to automatically stop the instance?

Hrm i dont think we should do that. Just prevent subsequent starts.

File: lxd/instance/drivers/driver_lxc.go [hunk 3]
```diff
@@ -2582,6 +2582,11 @@ func (d *lxc) validateStartup(statusCode api.StatusCode) error {
 		return fmt.Errorf("The image used by this instance requires nesting. Please set security.nesting=true on the instance")
 	}
 
+	// Check if instance is start protected.
+	if shared.IsTrue(d.expandedConfig["security.protection.start"]) {
+		return fmt.Errorf("Instance is protected")
```
Comment (tomponline) [inline lxd/instance/drivers/driver_lxc.go:2587]: Please can we change this to 

> Instance is protected from being started

Comment (qianzhangxa) [reply lxd/instance/drivers/driver_lxc.go:2587]: Sure. And do we need to change [this line](https://github.com/canonical/lxd/blob/5916f950f9afb84a967f9e67b724723d5d8ec6c4/lxd/instance/drivers/driver_qemu.go#L6115) too? Maybe in a different PR?

Comment (tomponline) [reply lxd/instance/drivers/driver_lxc.go:2587]: ah yeah, you can do it in the same PR just a different commit please, good spotting.

**Comment by qianzhangxa** [comment]
> > @tomponline I am thinking about this case: If we set `security.protection.start` to `true` for a **running** instance, do we need to automatically stop the instance?
> 
> Hrm i dont think we should do that. Just prevent subsequent starts.

Agree. The reason that I asked this question is that we may run into this case: A running instance with `security.protection.start` set to `true` which is a bit weird, but I guess it should be OK since that config option is just for preventing **subsequent** starts.

File: lxd/instance/drivers/driver_lxc.go (hunk 3 above)
Comment (qianzhangxa) [reply lxd/instance/drivers/driver_lxc.go]: Done

File: lxd/instance/drivers/driver_qemu.go [hunk 4]
```diff
@@ -1095,6 +1095,11 @@ func (d *qemu) validateStartup(stateful bool, statusCode api.StatusCode) error {
 		return fmt.Errorf("Stateful start requires migration.stateful to be set to true")
 	}
 
+	// Check if instance is start protected.
+	if shared.IsTrue(d.expandedConfig["security.protection.start"]) {
+		return fmt.Errorf("Instance is protected")
```
[the remark this answers is not in the document]
Comment (qianzhangxa) [reply lxd/instance/drivers/driver_qemu.go]: Done

**Review by tomponline — DISMISSED** [review]
Thanks!

**Comment by tomponline** [comment]
Ah please can you sign your commits

https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#ssh-commit-signature-verification

**Comment by qianzhangxa** [comment]
> Ah please can you sign your commits
> 
> https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#ssh-commit-signature-verification

Done :)

**Review by ru-fu — COMMENTED** [review]
The documentation update looks good, just two small nitpicks.

File: doc/howto/instances_manage.md [hunk 5]
```diff
@@ -131,6 +131,11 @@ Once it is running, you can select the {guilabel}`Terminal` tab to access the in
 ```
 ````
 
+### Prevent accidental start of instances
+
+- To protect a specific instance from being started, set {config:option}`instance-security:security.protection.start` to `true` for the instance.
+  See {ref}`instances-configure` for instructions.
```
[the remark this answers is not in the document]
Comment (qianzhangxa) [reply doc/howto/instances_manage.md]: Done

File: lxd/instance/instancetype/instance.go [hunk 6]
```diff
@@ -311,6 +311,15 @@ var InstanceConfigKeysAny = map[string]func(value string) error{
 	//  shortdesc: Prevents the instance from being deleted
 	"security.protection.delete": validate.Optional(validate.IsBool),
 
+	// lxdmeta:generate(entities=instance; group=security; key=security.protection.start)
+	//
+	// ---
+	//  type: bool
+	//  defaultdesc: `false`
+	//  liveupdate: yes
+	//  shortdesc: Prevents the instance from being started
```
[the remark this answers is not in the document]
Comment (qianzhangxa) [reply lxd/instance/instancetype/instance.go]: Done. And I updated the short description for `security.protection.delete` as well in a [new commit](https://github.com/canonical/lxd/pull/13824/commits/d629e126d1650708996d4610eedeca59a5ccf09c).

**Review by ru-fu — COMMENTED** [review]
Can you update the metadata again?

**Comment by qianzhangxa** [comment]
> Can you update the metadata again?

Done.

**Review by ru-fu — APPROVED** [review]
Thanks! The docs look good now. :)

**Comment by tomponline** [comment]
thanks!

**Pull request merged**